### ISSUE TITLE
Checking for existence in std::map - count vs find

### ISSUE THREAD

**Issue by SunBlack** [PointCloudLibrary/pcl#4083]
Currently we are using a lot of iterators just to check if an element exists in a map, like here:
https://github.com/PointCloudLibrary/pcl/blob/f9f214f34a38d5bb67441140703a681c5d299906/visualization/src/pcl_visualizer.cpp#L3261-L3268

To increase readability we have two options:
Option 1) Use `auto`
```
  auto am_it = cloud_actor_map_->find (id);
  if (am_it != cloud_actor_map_->end ())
  ...
```
Option 2) As single line:
```
  if (cloud_actor_map_->find (id) != cloud_actor_map_->end ())
  ...
```
Option 3) With `count`
```
  if (cloud_actor_map_->count (id) > 0)
  ...
```

See similar discussion on [StackOverflow](https://stackoverflow.com/questions/25490357/checking-for-existence-in-stdmap-count-vs-find).

Pro/Cons:
- Option 1) iterator variable is longer valid as necessary
- Option 2) In case of a long variable name we may reach the maximum line length defined by clang-format => line break
- Option 3) Shortest and imho best to read, but performance maybe be a few CPU ticks slower (GCC should be equal, but MSVC not sure)

So prefer performance or readability?

**Comment by larshg** [comment]
Option 3, the link of complexity is to size of the map, not count.

This is [count](https://en.cppreference.com/w/cpp/container/map/count).

Which is: Logarithmic in the size of the container.

**Comment by SunBlack** [comment]
You are right, happens when some libraries are using `count()` as equivalent to `size()` (alias Qt ;-) )

**Comment by larshg** [comment]
Hehe, why cant people stick to the same names :-D

As for the options I'd go for option 3, since find and count has same complexity. But the count makes the reason more clearer. Ie. Is the key we are about to add, already present.

**Comment by kunaltyagi** [comment]
* If it were not a map, I'd not go for option 3 (increased complexity otherwise)
* If we're using the iterator somewhere, I'd not go for option 2 and 3 (double work otherwise)
* If none of these apply, I'd go for option 3

Since you explicitly stated last point stands, I also vote for option 3

**Comment by SergioRAgostinho** [comment]
3) Thank you for taking the time to look into this.

**Comment by Morwenn** [comment]
4. Wait for C++20 and switch to `.contains()` then.

Given the explanation 3. seems like a sensible option.

**Comment by gnawme** [comment]
For the time being, assuming you're using C++17, use _if-with-initializer_ syntax:

`if (auto am_it = cloud_actor_map_->find (id); am_it != cloud_actor_map.end()) {...}`

**Comment by kunaltyagi** [comment]
C++14 only, at least for the moment.

**Comment by SergioRAgostinho** [comment]
The decision for selecting the minimum C++ standard we support was based on which mode comes enabled by default in the most popular current version of the compilers I believe. Example: clang switched to c++14 in [version 6](https://releases.llvm.org/6.0.0/tools/clang/docs/ReleaseNotes.html#c-language-changes-in-clang). I haven't tracked what's the current default dialect for the gcc/clang/cl combo.

**Comment by kunaltyagi** [comment]
All latest compiler still have C++14 as the default dialect. I was going for the [supported dialect](https://en.cppreference.com/w/cpp/compiler_support), because C++14->C++17 is a breaking change in some cases and I have a feeling most compilers will jump from C++14 to C++23 (17 isn't novel enough, 20 is incomplete), similar to C++98 to C++14.

**Comment by SergioRAgostinho** [comment]
We can go with that. There's no need to make our lives harder in trying to remain C++14 compatible given our limited manpower. Beware that nvcc support for C++17 is non-existent at the moment, but I assume that problem can be sidetracked once we have access to CUDA as a language from within CMake (likely to happen in the same update to 18.04 as baseline).

**Comment by kunaltyagi** [comment]
> Beware that nvcc support for C++17 is non-existent

That's true. Hopefully we don't need c++17 features for device code (not that we've needed them so far 😄)

> dialect for the gcc/clang/cl combo

Min versions with C++17 support
* GCC: 7 (18.04)
* Clang: 6 (16.04)
* MSVC: 19.14 (VS 2017 15.7)
* nvcc: :rofl: 11 (unreleased, also comes with `--allow-unsupported-compiler` which if not dropped will allow the co-pilot compiler to move ahead, like Clang 10)

**Comment by gnawme** [comment]
> For the time being, assuming you're using C++17, use _if-with-initializer_ syntax:
> 
> `if (auto am_it = cloud_actor_map_->find (id); am_it != cloud_actor_map.end()) {...}`

@kunaltyagi Are you good with this suggestion?

Is there a ticket to convert the `CMakeLists.txt` files to use C++17? You would also have to set this for GCC:

`set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++1z")`
